torch/utils/data/sampler.py

```diff
@@ -163,8 +163,12 @@ def __iter__(self) -> Iterator[int]:
             yield from torch.randint(high=n, size=(self.num_samples % 32,), dtype=torch.int64, generator=generator).tolist()
         else:
             for _ in range(self.num_samples // n):
-                yield from torch.randperm(n, generator=generator).tolist()
-            yield from torch.randperm(n, generator=generator).tolist()[:self.num_samples % n]
+                indices = torch.randperm(n, generator=generator)
+                for i in indices:
+                    yield i.item()
```

Comment:
one minor issue is that repeated `.item()` are slow: https://github.com/pytorch/pytorch/issues/29973

so maybe worth instead `yield from indices.tolist()`?

Reply:
Thanks for the suggestion! I've updated the code to call `.numpy` on the tensor before iterating on it, which should avoid the slow `.item()` calls

Reply:
Well, if you can afford an extra allocation, why not just `yield from indices.tolist()`? Because the indices take too much memory? 

I don't know what's the current state of affaires on obligation of numpy dependency

Reply:
Maybe worth filing a feature request to iterate Python items?  Or supporting memoryview on tensors (so that it can be iterated)...

Reply:
I've added some comments in the PR description, but the main issue is that calling `.tolist()` on torch tensor of size 1billion+ adds a massive garbage collection overhead because we just allocated billions of individual python int objects that need to be managed separately. By using a numpy array instead, the garbage collector only needs to keep track of 1 objects regardless of the dataset size, making garbage collection much faster during training.

Reply:
Please upvote https://github.com/pytorch/pytorch/issues/43949 :) then it could be iterated without numpy

Reply:
i created also https://github.com/pytorch/pytorch/issues/103352

Reply:
could this just be `yield from torch.randperm(n, generator=generator).numpy()` (+ some indexing) keeping the existing terse syntax?

Reply:
Done in 7a7a1e96e09a73b48e892352836444926ef5ba4a

Reply:
@aviverma01 Should we also change the other branch, `yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=generator).tolist()` ?

Reply:
@aviverma01 Should we also change the other branch, `yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=generator).tolist()` ?

Reply:
@aviverma01 Should we also change the other branch, `yield from torch.randint(high=n, size=(32,), dtype=torch.int64, generator=generator).tolist()` ?

Reply:
Done in b450690464b